### PR TITLE
PBI #76851 Fix warning  React Hook useEffect has a missing dependency

### PR DESCRIPTION
**Pull request by ljairath**
warning  React Hook useEffect has a missing dependency: 'tabs'. Either include it or remove the dependency array react-hooks/exhaustive-deps

# Summary of changes
Memoizes Tabs using useMemo. 
Reference: https://kinsta.com/knowledgebase/react-hook-useeffect-has-a-missing-dependency/ 

## Frontend
List all significant changes to Gatsby code

### Memoize Tabs
const tabs = useMemo(() => {
    return props.pageData?.relationships?.field_tabs ?? []
  }, [props.pageData?.relationships?.field_tabs])

### Added Tabs array in useEffect dependency
  }, [tabs])

[ x] My changes are accessible (at minimum WCAG 2.0 Level AA)
[ x] My changes are responsive and appear as expected on mobile and desktop views.
[ x] After merging, I will update the [Content Hub documentation](https://uoguelphca.sharepoint.com/sites/UniversityContentHubInformationGroup) and ensure the Content Hub trainer understands how my changes will affect users.

# Test Plan
Run gatsby develop/build on gus
Notice the warning has disappeared

### CODE REVIEW

**Comment by RazaMM** [comment]
Is the useMemo necessary to get rid of the error? useMemo really should only be used for performance heavy calculations as it has overhead associated with it, but all we are doing is setting a variable, so I think it would be better to not use it.

**Review by RazaMM — CHANGES_REQUESTED** [review]
See previous comment

**Comment by tqureshi-uog** [comment]
Ran a build with the latest changes (including the removal of `useMemo`) and the warning is gone. @ljairath For future reference, it helps to run a warning through https://copilot.microsoft.com (make sure you switch to "More precise" mode first) and see what it recommends. While it's not always correct, in this case, it recommended the cleanest and simplest solution, i.e. no change to the original code necessary *except* adding `tabs` to the previous empty square brackets on line 32 (i.e. the dependency array):

```
useEffect(() => {
  const params = new URLSearchParams(window.location.search);
  const titleOrID = params.get("tab");

  if (titleOrID) {
    // Try to find a tab with a matching drupal id
    let found = tabs.find((tab) => tab.drupal_id === titleOrID);

    // If we couldn't find a matching id, then try and find a matching title
    found ??= tabs.find((tab) => titleOrID === slugify(tab.field_tab_title.toLowerCase()));

    // Set the active tab to whichever was found
    // If neither could be found then set the active tab to the first tab.
    setActiveTab(found?.drupal_id ?? tabs[0]?.drupal_id);

    // Scroll into view if a tab was found
    found && container.current?.scrollIntoView({ block: "start", inline: "nearest" });
  }
}, [tabs]); // Dependency array includes 'tabs'

**Review by tqureshi-uog — APPROVED** [review]
I notice Raza's fix also tidied up some null checks, which copilot didn't do because it wasn't asked. The copilot code would have still been an effective starting point, however, particularly if one is new to React.

**Comment by ljairath** [comment]
Hi Tanya.

Good morning. Great suggestion. I appreciate it very much!

Maybe it's beside the point now... however, I tried with [tabs], but the build gave me another error. I don't remember the exact error, but the error message suggested using the useMemo because tabs were an object/array and not a simple variable.

Thanks again!
Lalit


________________________________

**Comment by tqureshi-uog** [comment]
> Maybe it's beside the point now... however, I tried with [tabs], but the build gave me another error. I don't remember the exact error, but the error message suggested using the useMemo because tabs were an object/array and not a simple variable. 

I never saw any warning to that effect when I ran my own builds to test this PR. I also checked with Copilot again and it agrees with what Raza said earlier:

"Using useMemo to get rid of the missing dependency warning in a useEffect hook might not be the best approach. The purpose of useMemo is to memoize values, not to bypass dependency warnings. The warning is there to help you identify potential bugs related to stale closures and missing dependencies.

If an earlier version of your code used useMemo to define a value that was then used inside a useEffect, and this was done to avoid including a dependency in the useEffect’s dependency array, it was likely a misuse of useMemo. The correct approach is to include all relevant dependencies in the dependency array of useEffect.

Here’s an example of a potential misuse:

```
// Misuse of useMemo to avoid dependency warnings
const memoizedTabs = useMemo(() => tabs, [tabs]);

useEffect(() => {
  // ... effect logic using memoizedTabs ...
}, []); // Empty dependency array, missing memoizedTabs
```

In this case, memoizedTabs is unnecessary because it’s simply a memoized version of tabs, and tabs should be included in the dependency array directly:

```
useEffect(() => {
  // ... effect logic using tabs ...
}, [tabs]); // Correctly include tabs in the dependency array
```

The useEffect hook should have a complete list of dependencies to ensure it behaves correctly. If you’re seeing a warning about a missing dependency, it’s usually a sign that you need to rethink how you’re structuring your effect and what dependencies it truly has.

In summary, useMemo should be used for its intended purpose of optimizing performance by memoizing expensive calculations, not for avoiding React Hook dependency warnings. Always include all used values in the dependency array of useEffect to follow React’s best practices and avoid potential issues."

@RazaMM would you say this is good to merge now?

**Comment by RazaMM** [comment]
@tqureshi-uog Yeah I think its good to merge now. 

Just to comment on what @ljairath was talking about, I tried (and im assuming this is what he did as well) just adding tabs to the dependency array, and theres the following warning:

`The 'tabs' logical expression could make the dependencies of useEffect Hook (at line 32) change on every render. To fix this, wrap the initialization of 'tabs' in its own useMemo()
Hook`. 

The reason was that originally we had the following `const tabs = props.pageData?.relationships?.field_tabs ?? []`. The problem with that is that if `props.pageData?.relationships?.field_tabs` is null we set tabs equal to a empty array, useEffect sees that tabs changed and causes a rerender, on that rerender tabs would be set to a brand new empty array. This would happen on every render which is unnecessary and why react suggests useMemo which only gets calculated as needed and not on every render. Although as you said useMemo is really only needed for expensive calculations, not something like this.

So I got around useMemo by just getting rid of the need for `props.pageData?.relationships?.field_tabs ?? []` and use null checks (with the optional chaining operator) instead.

**Review by RazaMM — APPROVED** [review]
